forte/orbital-helpers/aosubspace.cc

```diff
@@ -62,84 +65,74 @@ std::vector<std::string> mysplit(const std::string& input, const std::string& re
+        subspace_str.push_back(s);
+    }
+
+    // Create a basis set parser object and read the minimal basis
+    std::shared_ptr<psi::Molecule> molecule = wfn->molecule();
+    std::shared_ptr<psi::BasisSet> min_basis = wfn->get_basisset("MINAO_BASIS");
+
+    // Parse the atom normals used to project the p orbitals
+    std::map<std::pair<int, int>, psi::Vector3> atom_to_normal;
+    for (const auto& item : atom_normals) {
+        auto atom = item.first.cast<std::pair<int, int>>();
+        auto normal = item.second.cast<std::array<double, 3>>();
+        atom_to_normal[atom] = psi::Vector3(normal);
+    }
+
+    // Create an AOSubspace object
+    AOSubspace aosub(subspace_str, molecule, min_basis, atom_to_normal);
+
+    // build the projector
+    Ps = aosub.build_projector(wfn->basisset());
+
+    // print the overlap of the projector
+    psi::SharedMatrix CPsC = Ps->clone();
+    CPsC->transform(wfn->Ca());
+    double print_threshold = 1.0e-3;
+    auto irrep_labels = molecule->irrep_labels();
+    outfile->Printf("\n  ==> Orbital Overlap with AO Subspace (> %.2e) <==\n", print_threshold);
+    outfile->Printf("\n    =======================");
+    outfile->Printf("\n    Irrep   MO  <phi|P|phi>");
+    outfile->Printf("\n    -----------------------");
+    for (int h = 0, nirrep = CPsC->nirrep(); h < nirrep; h++) {
+        for (int i = 0, size = CPsC->rowspi(h); i < size; i++) {
+            if (CPsC->get(h, i, i) > print_threshold) {
+                outfile->Printf("\n    %4s  %4d  %10.6f", irrep_labels[h].c_str(), i + 1,
+                                CPsC->get(h, i, i));
             }
         }
-        outfile->Printf("    ========================\n");
     }
+    outfile->Printf("\n    ========================");
     return Ps;
 }
 
-AOSubspace::AOSubspace(std::shared_ptr<Molecule> molecule, std::shared_ptr<BasisSet> basis)
-    : molecule_(molecule), basis_(basis) {
-    startup();
-}
-
 AOSubspace::AOSubspace(std::vector<std::string> subspace_str,
                        std::shared_ptr<psi::Molecule> molecule,
-                       std::shared_ptr<psi::BasisSet> basis)
-    : subspace_str_(subspace_str), molecule_(molecule), basis_(basis) {
+                       std::shared_ptr<psi::BasisSet> basis,
+                       std::map<std::pair<int, int>, psi::Vector3> atom_normals, bool debug_mode)
+    : subspace_str_(subspace_str), molecule_(molecule), min_basis_(basis), subspace_counter_(0),
+      atom_normals_(atom_normals), debug_(debug_mode) {
     startup();
 }
 
-void AOSubspace::find_subspace() {
-    parse_basis_set();
-    parse_subspace();
-}
-
 void AOSubspace::startup() {
-    //    outfile->Printf("  ---------------------------------------\n");
-    //    outfile->Printf("    Atomic Orbital Subspace\n");
-    //    outfile->Printf("    written by Francesco A. Evangelista\n");
-    //    outfile->Printf("  ---------------------------------------\n");
-
```

Comment:
I wonder `lm_labels_cartesian_ ` should be removed because it is not used. We could test the `min_basis_` to make sure it is `pure_am`.

Reply:
I suppose the idea was to support both 5D and 6D type of functions, but that never happened. So I think it's a good idea to remove cartesians and check for pure_am. I don't remember, but are pure_am basis sets still pure cartesians, like px, py, pz?

Reply:
I think `has_puream` is `false` if using `cartesian` in the basis but `true` for `spherical`. I do not like the ordering of `lm_labels_spherical_` because I have to consider the mismatch between it (pz, px, py) and the normal vector (nx, ny, nz).